### DISCUSSION TITLE
What is a practical use case for getServerSideProps?

### DISCUSSION THREAD

**Discussion by gopinav** [vercel/next.js#24974] [Help]
Hello, I'm currently learning about pre-rendering in Next JS. Although Next JS provides decent "scenarios" for using SSG/ISR and SSR, I can't seem to pin the usage of getSterverSideProps to a practical usage. I understand it is needed for when we need the request context, for pre-rendering etc but if you could help me with a use case you are currently using it for, would be really helpful. 

For example,

1. getStaticProps can be used for marketing sites and blog sites
2. getStaticProps and getStaticPaths can be used for example in an e-commerce site
3. ISR can be used for example, in a documentation site
4. But when would I use getServerSideProps?  (A user dashboard might make sense but client side rendering suffices with the only drawback of seeing a loading screen since SEO is not important?)

Thanks in advance

**Comment by hexionaut** [comment]
Hello,

You will use getServerSideProps (so SSR) when you have data that change really often. This allows your data to always be up to date.

**Reply by gopinav** [reply]
Hello,

I understand that but I'm looking for a practical use case.

**Reply by leodr** [reply]
Twitter for example. Tweets need to be up to date and content is per-user, but the site should also have great SEO. So you need server-side rendering via `getServerSideProps`.

**Reply by gopinav** [reply]
Hello, would you be able to elaborate how the SEO for twitter would work? I'm guessing a user's feed might not be SEO friendly? What content of the page are we looking to optimize for search engines?

**Reply by leodr** [reply]
Pretty much every tweet is publicly accessible with its own URL. It can also have comments, so it needs to be up to date to users and optimized for search engine crawlers.

**Reply by lkbr** [reply]
Couldn't `getStaticProps` with a low `revalidate` and `fallback: blocking` handle this?

**Reply by leodr** [reply]
Not exactly. `getStaticProps` only works if the content is the same for every visitor. But on Twitter, you can block people for example, so their Tweets would not be shown for you, but for everybody else visiting the page.

`getServerSideProps` is the only way with Next.js to do server-side authentication and personalized content (server-side rendered, you can obviously do personalized content client-side rendered).

**Reply by gopinav** [reply]
I see. So twitter feed is not for a user only data but rather 'personalised' for a user. Because of that, the tweet should still be SEO friendly as it is visible pretty much for anyone. Would it be ok to say that?

**Reply by leodr** [reply]
Yeah, that pretty much sums it up. The content is personalized but also critical for SEO, so it can't be static.

**Reply by lkbr** [reply]
Ah, yes, thank for elaborating, I hadn't considered personalisation + SEO.

**Comment by jportela** [comment]
One use case for SSR is if you have very different app skeletons, that depend on user specific data.

For example, if you want to show a landing or login page for unauthenticated users, but want to show a dashboard or a specific page for the user, and don't want the user to see a "jump" between the unauthenticated page and their page when the page is first rendered.

Also, building a great client-side experience (that adds personalized user content via dynamic http requests) is harder to do properly (with loading states and great UX), than just loading all the user data you need via `getServerSideProps`, and have it already there on first render.

Another one could be customization, for example, if you are running A/B tests or feature flags on your web app (that are different depending on the user), the page might look very different between users, and as such, you might want to use `getServerSideProps` to avoid "jumps" in the functionality you offer.

**Reply by gopinav** [reply]
Would it be fair to say that the 'jump' or 'loading screen' is the main problem you're trying to solve in your examples?

**Reply by jportela** [reply]
Yes, the tradeoff is that while the server is processing the request, the user will see a browser loading state instead, which will slow down everything either way. Newer frontend patterns favor SSG + dynamic HTTP loading, since you can deliver a faster and more personalized loading experience (ie: skeleton states, loading indicators).

**Reply by gopinav** [reply]
Thank you for explanation

**Comment by lkbr** [comment]
The only case I've found to choose `getServerSideProps` over `getStaticProps` so far is a unique requirement of query params + SEO. Typically things like product list pages with complex filtering that are synced to query params for SEO or so the user can share the state of their filtering. Think of an Amazon category page that has pagination, filtering and sorting. 

We're using catch all routes to handle hierarchal categories e.g. website.com/products/electronics/phones and query params to handle filtering e.g website.com/products/electronics/phones?brand=apple&page=2.

You could use getStaticProps for this, but because it's at build time its unaware of the brand and page params. You could use useSWR to get the right data once everything has loaded, but the initial load would not include the data you want for SEO. 

You could try to avoid query params and rely on the url structure e.g. website.com/products/electronics/phones/apple/2 but filtering can get complex and you're relying on param order to determine what each param represents. 

We host on Vercel so we're using https://vercel.com/docs/edge-network/caching#stale-while-revalidate which retains most of the benefits of `getStaticProps`. 

We use and love `getStaticProps` for every other kind of page. 

Edit: I think I should also mention we only use getServerSideProps for the initial load, we then use useSWR (with initialState from passed from SSR) to get new data as the user is filtering and then we use shallow routing to push the filtering state to the url, so if they share the url or refresh, SSR will kick in and the process will repeat.

**Reply by gopinav** [reply]
That's a fantastic example! Thank you for elaborating on the post initial load process as well. I'm just wondering, from an SEO perspective, how do query parameters affect SEO? Do filters have such an impact on SEO that you'd use SSR instead of client side fetching? Or is it just the ease of coding and getting data at once rather than fetching the product page, and then fetching again client side based on query params?

**Reply by lkbr** [reply]
You probably know this stuff, but for the sake of completeness.

Nothing special about query params SEO-wise, we just need a way to house distinct pages for different combinations of search terms we'd like to rank for. If your filtering is simple enough you could map your filters to structured urls and use getStaticProps (website.com/products/filter1/filter2/filter3/pagination) but I've found after 3+ combinations this gets unwieldy so we resort to query params. 

Consider this filter: website.com/products/electronics/phones?brand=apple&location=cambridge&page=2

SEO requirements are closely related with message/social media sharing requirements e.g. a user might want to share a page with the results they have filtered. For a user, it might be okay for getStaticProps to render all phones and then after first load when router.isReady is true for useSWR to fetch page 2 of apple + cambridge phones. But for SEO, the html from the first load is important for search engine crawling/indexing (its a huge topic to read about). Using the strategy so far, to google, these would probably all appear to have the same content:

- website.com/products/electronics/phones?brand=apple&location=cambridge&page=2
- website.com/products/electronics/phones?brand=samsung
- website.com/products/electronics/phones?brand=samsung&location=manchester&page=9

Where as using the query params passed by `getServerSideProps` we can determine exactly what to fetch and then pass it to the initialData of useSWR. Then use caching to alleviate the reason against getServerSideProps.

> Or is it just the ease of coding and getting data at once rather than fetching the product page, and then fetching again client side based on query params

This hasn't been reason for us yet, since we're only doing two fetches on these kinds of pages 1. The category (title, description, image etc) from our db and 2. the list of products from algolia. But I could see how SSR could help with more complex waterfall fetching.

**Reply by gopinav** [reply]
Thanks again for the detailed explanation.
